modules/gdscript/gdscript_formatter.cpp

```diff
@@ -0,0 +1,231 @@
+/*                      https://godotengine.org                          */
+/*************************************************************************/
+/* Copyright (c) 2007-2019 Juan Linietsky, Ariel Manzur.                 */
+/* Copyright (c) 2014-2019 Godot Engine contributors (cf. AUTHORS.md)    */
+/*                                                                       */
+/* Permission is hereby granted, free of charge, to any person obtaining */
+/* a copy of this software and associated documentation files (the       */
+/* "Software"), to deal in the Software without restriction, including   */
+/* without limitation the rights to use, copy, modify, merge, publish,   */
+/* distribute, sublicense, and/or sell copies of the Software, and to    */
+/* permit persons to whom the Software is furnished to do so, subject to */
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "gdscript_formatter.h"
+
+namespace {
+
+bool is_end_of_expression(GDScriptTokenizer::Token token) {
+	switch (token) {
+		// Any identifier or constant, built-in or otherwise, indicates an expression
+		// A closing bracket, curly brace, or parenthesis also indicates an expression
+		case GDScriptTokenizer::TK_IDENTIFIER:
+		case GDScriptTokenizer::TK_CONSTANT:
+		case GDScriptTokenizer::TK_SELF:
+		case GDScriptTokenizer::TK_BUILT_IN_TYPE:
+		case GDScriptTokenizer::TK_BUILT_IN_FUNC:
+		case GDScriptTokenizer::TK_BRACKET_CLOSE:
+		case GDScriptTokenizer::TK_CURLY_BRACKET_CLOSE:
+		case GDScriptTokenizer::TK_PARENTHESIS_CLOSE:
+		case GDScriptTokenizer::TK_CONST_PI:
+		case GDScriptTokenizer::TK_CONST_TAU:
+		case GDScriptTokenizer::TK_WILDCARD:
+		case GDScriptTokenizer::TK_CONST_INF:
+		case GDScriptTokenizer::TK_CONST_NAN:
+			return true;
+		default:
+			// Anything else is a separator of some kind
+			return false;
+	}
+}
+
+bool should_insert_space(GDScriptTokenizer::Token left, GDScriptTokenizer::Token cur, GDScriptTokenizer::Token right) {
```

Comment:
Method parameters should all be prefixed with `p_`, unless they are references modified by the method, in which case they start with `r_`. So most prototypes need to be adjusted to use `p_`.